### PR TITLE
I18n tools

### LINKED ISSUE #7709: kolibri-tools needs improved message extraction for Kolibri plugins
Per the last comments on: https://github.com/learningequality/kolibri/pull/7671

1. We need to update the `kolibri-tools` package to properly handle the plugin paths in Kolibri. The PR was merged and `kolibri-tools` subsequently release to accommodate the changes in that PR which make the i18n tools available cross-product.

2. Context transfer has hard-coded kolibri paths. Should require relatively light refactor to accommodate Studio as well

### LINKED ISSUE #8218: String context transfer from Crowdin fails for Vue files
## Observed behavior
@nucleogenesis did some work to fix this in #8048, but it might need some retouches.

Running `make i18n-download` apparently executes correctly, reports context as transferred, and errors out only at the last step (linter). However, when I check status, it seems as if only `PO`, `MO` and `JSON` and 2 `JS` files are modified, and not a single `Vue` file, which is where most of our strings and context should be.

## Errors and logs
Terminal output: 
[make i18n-download.txt](https://github.com/learningequality/kolibri/files/6850916/make.i18n-download.txt)

## Expected behavior
All the Vue files have the added context transferred from Crowdin.

## User-facing consequences
...

## Steps to reproduce
…

## Context
…

cc @rtibbles

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary
* Cleans up and consolidates i18n code
* Make CSV to JSON conversion create separate JSON for each bundle
* Remove hard coding of project and path information
* Makes CLI more user friendly
* Makes JS CLI commands within kolibri-tools for all i18n code
* Consolidate language_info.json into a single file that gets packaged for export at publish
* Consolidates all AST parsing to use the babel parser, and uses consistent functions for detection of translation functionality
* Fixes context transfer code to pick up context from any downloaded CSV file
* Transfers context from Crowdin into source code
* Updates JSON message files based on currently used strings
* Adds an `i18n` entry to the `setup.cfg` to have shared configuration across Python and JS i18n tooling

## References
Fixes #7709
Fixes #8218

## Reviewer guidance
Confirm that different translation steps still function as expected.
Do the changes in the JSON file change commit make sense?
Is the context transfer complete (last commit)?
Does the breakdown of JS CLI commands make sense?
Do any docs needs to be updated? I think the workflow is still the same.

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: packages/kolibri-tools/package.json
Comment (indirectlylit) [inline packages/kolibri-tools/package.json:59]: Doing this would avoid adding yet another copy of lodash:

```suggestion
    "lodash": "^4.17.20",
```

ref: https://github.com/learningequality/kolibri/pull/8224/files#diff-51e4f558fae534656963876761c95b83b6ef5da5103c4adef6768219ed76c2de

**Comment by radinamatic** [comment]
A part from the font generation error already mentioned in Slack, I am unable to see the latest string context changes pulled from the Crowdin when running the `make i18n-download branch=release-v0.14.x` locally on this branch.

For example, I've changed the context of the [`CoachClassListPage.noClassesDetailsForAdmin`](https://crowdin.com/translate/kolibri/3810/en-es#276404) string, project has been built successfully, CSVs downloaded but the context change is not being updated in the local CSV.

Cowdin | local CSV
--- | ---
Selection_022 | LEDev2104 (start)  Running  - Oracle VM VirtualBox_019

Terminal also reports one Vue file skipped:

```
Kolibri: Skipping /home/didivm/LE/kolibri/kolibri/plugins/coach/assets/src/views/common/LearnerExerciseReport.vue because $trs property was given an empty object
```

File: setup.cfg
Comment (indirectlylit) [inline setup.cfg]: could do something along the lines of...


```suggestion
[kolibri:i18n]
project = kolibri
locale_data_folder = kolibri/locale
```

where eventually we'd also have e.g.

```ini
[kolibri:i18n]
project = studio
locale_data_folder = contentcuration/locale
```

**Comment by rtibbles** [comment]
Does the context make it into the source code? If Crowdin is not putting it into the CSV files, not sure what I can do about that - but it doesn't seem to be predictable which CSV files the context ends up in on export.


> Kolibri: Skipping /home/didivm/LE/kolibri/kolibri/plugins/coach/assets/src/views/common/LearnerExerciseReport.vue because $trs property was given an empty object

This is intentional, there are no strings, so it is skipping. This is more a warning of a pointless $trs object in the component than anything else.

**Comment by rtibbles** [comment]
I reran the context generation, it also did not reflect the update in Crowdin.

The issue seems to be that the old context is present in some of the downloaded CSV files, while the new context is only present in one. I don't think our code has any way to decide which is the newest one.

**Comment by radinamatic** [comment]
> Does the context make it into the source code?

If I understand correctly this is the Vue file where the context should have been updated, and no, it does not seem to make it in there:

LEDev2104 (start)  Running  - Oracle VM VirtualBox_020

I don't know the details how the command is updating the context when the CSV files are downloaded. Could it be we are not fetching correctly from Crowdin, or it needs more time (?) to update itself... 😕 

The most recent context is saved in the **source** CSV files, that can be downloaded manually from the **Settings > Files** page, and I can also download the CSV file in question (`kolibri.plugins.coach.app-messages.csv`) from the Crowdin **Editor** page in all the languages, both display the new context correctly:

[kolibri.plugins.coach.app-messages.csv](https://github.com/learningequality/kolibri/files/6889405/kolibri.plugins.coach.app-messages.csv)

However, when downloaded manually from the **Settings > Translations** page, the same CVS file inside any language folder of the ZIP, does not have the updated context.

**Comment by radinamatic** [comment]
A part from the context issue, devserver builds & runs OK, displays locales correctly and I did not notice any other apparent problems. We need to have the new 0.15.x branch on Crowdin to translate the new strings for S2S, so the context is not a blocker at this moment. 

I'm going to approve this and if @indirectlylit does not have other suggestions, I vote we merge and come back to figure out the context update issue in subsequent PRs.

**Review by radinamatic — APPROVED** [review]
So great seeing all this context for user strings finally makes it into the code! 💯 
Thank you @rtibbles, and thank you @dufcrule!

**Comment by indirectlylit** [comment]
>  if @indirectlylit does not have other suggestions

Just beginning my review now. If there's urgency to merge I can always file follow-up issues

File: packages/kolibri-tools/build_kolibri_tools.js
Comment (indirectlylit) [inline packages/kolibri-tools/build_kolibri_tools.js:38]: I thought that `kolibri-tools` was an independent package? This seems to still be coupled to it's current location in the Kolibri repo.

File: packages/kolibri-tools/lib/cli.js
Comment (indirectlylit) [inline packages/kolibri-tools/lib/cli.js]: Text in these options is hard to read when using `--help` in the terminal:

image

File: packages/kolibri-tools/lib/cli.js
Comment (indirectlylit) [inline packages/kolibri-tools/lib/cli.js]: With regard to `kolibri-tools i18n`, there are other more-specifically-named commands which makes this one a bit nebulous-looking:

```
Commands:
  build [options] <mode>      Build frontend assets for Kolibri frontend plugins
  lint [options] [files...]   Run linting on files or files matching glob patterns
  test [options]
  compress [files...]
  i18n-code-gen [options]
  i18n [options] <mode>
  i18n-csv-to-json [options]
  i18n-stats [options]
  i18n-profile [options]
```

Recommend choosing a more specific name and adding some more descriptions

image

File: packages/kolibri-tools/lib/cli.js
Comment (indirectlylit) [inline packages/kolibri-tools/lib/cli.js]: `-pf` is confusing because posix conventions would interpret it as `-p -f`

Names for `--` arguments are also usually lower-kebab-case rather than camel: `--plugin-file`. Similar note on other options in this file

File: packages/kolibri-tools/lib/cli.js
Comment (indirectlylit) [inline packages/kolibri-tools/lib/cli.js]: maybe cleaner as separate CLI commands with shared underlying code rather than a single command

File: packages/kolibri-tools/lib/i18n/SyncContext.js
Comment (indirectlylit) [inline packages/kolibri-tools/lib/i18n/SyncContext.js]: This predated your work, but mind expanding instances of `acc` to a more descriptive name?

File: packages/kolibri-tools/lib/cli.js
Comment (indirectlylit) [inline packages/kolibri-tools/lib/cli.js:22]: We haven't established this as a convention at LE, but I personally prefer to prefix internal module helper functions with an underscore

File: packages/kolibri-tools/lib/cli.js
Comment (indirectlylit) [inline packages/kolibri-tools/lib/cli.js]: general comment on CLI command names:

rather than naming them based on file formats (which are somewhat arbitrary) I think it would be long-term more helpful to name them based on intended goal.

For example, instead of `i18n-csv-to-json` we might call it `i18n-crowdin-to-git` or something along those lines. (I realize that might not be quite correct here, but I'm still processing the code)

File: packages/kolibri-tools/lib/i18n/csvToJSON.js
Comment (indirectlylit) [inline packages/kolibri-tools/lib/i18n/csvToJSON.js]: where ever these files end up, would be great to add a `readme.md` file that has a reference back to here for future i18n spelunkers

File: packages/kolibri-tools/lib/cli.js
Comment (indirectlylit) [inline packages/kolibri-tools/lib/cli.js]: Flagging that this is now the second meaning of "stats" in this file – the original being a mode of the `build` command. This makes searching through the file more confusing.

File: packages/kolibri-tools/lib/i18n/utils.py
Comment (indirectlylit) [inline packages/kolibri-tools/lib/i18n/utils.py:19]: somewhat brittle coupling to details of existing repo structure. Perhaps should be pulled from `setup.cfg` or some other global per-project config?

**Review by indirectlylit — COMMENTED** [review]
Great job – significant improvements here.

I haven't tested yet, just read through the commits. My comments below mostly pertain to longer-term concerns and nothing blocks merging to move forward with short-term updates.

I did not yet have a chance to test context transfers (cc @radinamatic) – if helpful I can take a look at this on Friday

File: packages/kolibri-tools/build_kolibri_tools.js
Comment (rtibbles) [reply packages/kolibri-tools/build_kolibri_tools.js:38]: This is the build tools - it is built inside the Kolibri repo, and makes a copy of this into the package, so that it is indeed independent of the Kolibri repo at publish time. Similarly to how we bundle up the core API spec to be accessible outside of the kolibri repo.

File: packages/kolibri-tools/lib/cli.js
Comment (rtibbles) [reply packages/kolibri-tools/lib/cli.js]: I am also happy to get rid of these options - I did not use the dry-run or dump options at all while testing.

File: packages/kolibri-tools/lib/cli.js
Comment (rtibbles) [reply packages/kolibri-tools/lib/cli.js]: I could also just separate this out into two separate commands - I was trying to follow the `build` command at first in putting all the modes under one roof, but I think it's simpler to just split this into two commands.

File: packages/kolibri-tools/lib/cli.js
Comment (rtibbles) [reply packages/kolibri-tools/lib/cli.js]: Quite a few popular Node CLI utilities use camelCase, which is why this was implemented like this previously.

See, e.g. jest: https://jestjs.io/docs/cli#options

(Webpack uses kebab-case, so it's not hard and fast)

File: packages/kolibri-tools/lib/cli.js
Comment (rtibbles) [reply packages/kolibri-tools/lib/cli.js]: I concur.

File: packages/kolibri-tools/lib/i18n/SyncContext.js
Comment (rtibbles) [reply packages/kolibri-tools/lib/i18n/SyncContext.js]: **Acc**knowledged.

File: packages/kolibri-tools/lib/cli.js
Comment (rtibbles) [reply packages/kolibri-tools/lib/cli.js:22]: I like that practice in Python, but less so in Javascript, where the definition of an internal function is based on whether it is exported.

File: packages/kolibri-tools/lib/i18n/utils.py
Comment (rtibbles) [reply packages/kolibri-tools/lib/i18n/utils.py:19]: I did think about adding it there - mostly this is a workaround for us only keeping one copy of `language_info.json` in the repository (previously it was duplicated in the locale folder and in this folder), this allows us to only replicate this default file at build time of the package.

The reason I decided not to add it as a global configuration was because it should only need to be specified when our language_info.json is insufficient.

File: setup.cfg
Comment (rtibbles) [reply setup.cfg]: Hrm, not a big fan of the `:`, but it seems like Python ini parsers that allow subsections (ConfigObj) do it in a completely different way to the ini file parser for nodejs that I'm using, so this might be simplest.

File: kolibri/core/assets/src/views/AppError/ReportErrorModal.vue
Comment (indirectlylit) [inline kolibri/core/assets/src/views/AppError/ReportErrorModal.vue]: looks like the previous change might have introduced an extra newline in the output?

**Review by indirectlylit — APPROVED** [review]
awesome, thanks for incorporating my feedback

File: kolibri/core/assets/src/views/AppError/ReportErrorModal.vue
Comment (rtibbles) [reply kolibri/core/assets/src/views/AppError/ReportErrorModal.vue]: I honestly have no idea why this happened - I hadn't changed anything in the code. One possibility is that the script is not idempotent. I'll take another look.

Comment (rtibbles) [reply kolibri/core/assets/src/views/AppError/ReportErrorModal.vue]: Hrm, it does seem to be idempotent.

Comment (rtibbles) [reply kolibri/core/assets/src/views/AppError/ReportErrorModal.vue]: OK, narrowed down the issue - seems to have been a problem that happened only when 'updating' context that hadn't changed. So I've just stopped it doing that, which seems more appropriate anyway!

**Pull request merged**
